File: 2022/day-9/day9.py

from typing import *
from math import copysign
from functools import partial
from operator import add, sub

Instruction = Tuple[str, int]
Position = Tuple[int, int]


def parse_line(line: str) -> Instruction:
    direction, amount = line.split()
    return direction, int(amount)


def modify(op: Callable[[int, int], int], head: Position, tail: Position) -> Position:
    hx, hy = head
    tx, ty = tail
    return op(hx, tx), op(hy, ty)


diff = partial(modify, sub)
move = partial(modify, add)


def clamp(a: int) -> int:
    return int(copysign(1, a)) if a else 0


def simulate(inst: Iterable[Instruction], length: int = 2):
    rope = [(0, 0)] * length
    visited = {rope[-1]}
    directions = dict(L=(-1, 0), R=(1, 0), U=(0, 1), D=(0, -1))
    for direction, amount in inst:
        vec = directions.get(direction)
        for _ in range(amount):
            rope[0] = move(rope[0], vec)
            for i in range(length-1):
                dx, dy = diff(rope[i], rope[i+1])  # distance between two segments
                if abs(dx) > 1 or abs(dy) > 1:  # should move tail?
                    # make sure move vector points along dx and dy
                    tail = rope[i+1] = move(rope[i+1], (clamp(dx), clamp(dy)))
                    if i == length - 2:  # tail of the rope as index i-2
                        visited |= {tail}
    return visited


with open('input.txt') as h:
    instructions = list(map(parse_line, h.read().splitlines()))
print('Part 1:', len(simulate(instructions)))
print('Part 2:', len(simulate(instructions, length=10)))

File: 2022/day-9/day9.rs
use std::collections::HashSet;

type Instructions = Vec<(char, i32)>;
type Point = (i32, i32);
type Rope = Vec<Point>;

fn parse_input(inp: &str) -> Instructions {
    inp.lines()
        .map(|l| {
            let mut parts = l.split(' ');
            let dir = parts.next().unwrap().chars().next().unwrap();
            let amount: i32 = parts.next().unwrap().parse().unwrap();
            (dir, amount)
        })
        .collect()
}

fn modify<T>(f: T, a: Point, b: Point) -> Point
where
    T: Fn(i32, i32) -> i32,
{
    let (ax, ay) = a;
    let (bx, by) = b;
    (f(ax, bx), f(ay, by))
}

fn add(a: Point, b: Point) -> Point {
    modify(std::ops::Add::<i32>::add, a, b)
}

fn diff(a: Point, b: Point) -> Point {
    modify(std::ops::Sub::<i32>::sub, a, b)
}

fn clamp(a: i32) -> i32 {
    if a == 0 {
        0
    } else {
        a.signum()
    }
}

fn simulate(instructions: &Instructions, length: usize) -> HashSet<Point> {
    if length < 2 {
        panic!("A rope must at least have a head and a tail (length > 2)")
    }
    let mut rope: Rope = (0..length).map(|_| (0, 0)).collect();
    let mut visited: HashSet<Point> = HashSet::new();
    visited.insert(*rope.last().unwrap());

    for &(direction, amount) in instructions {
        let vec: Point = match direction {
            'L' => (-1, 0),
            'R' => (1, 0),
            'U' => (0, 1),
            'D' => (0, -1),
            _ => panic!("Unknown direction {}", direction),
        };
        for _ in 0..amount {
            rope[0] = add(rope[0], vec);
            for i in 0..(rope.len() - 1) {
                let (dx, dy) = diff(rope[i], rope[i + 1]);
                if dx.abs() > 1 || dy.abs() > 1 {
                    rope[i + 1] = add(rope[i + 1], (clamp(dx), clamp(dy)));
                    if i == length - 2 {
                        visited.insert(rope[i + 1]);
                    }
                }
            }
        }
    }
    visited
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input");
    let instructions = parse_input(&contents);
    println!("Part 1: {:?}", simulate(&instructions, 2).len());
    println!("Part 2: {:?}", simulate(&instructions, 10).len());
}
